cajita/src/Cajita_Hilbert.hpp

```diff
@@ -0,0 +1,838 @@
+            {
+                int gval = 0;
+                int gtemp =
+                    ( w ^ ( w >> 1 ) ) ^ ( ( w + 1 ) ^ ( ( w + 1 ) >> 1 ) );
+                while ( gtemp >>= 1 )
+                    gval++;
+                wd = gval % 3;
+            }
+
+            vd = ( vd + wd + 1 ) % 3;
+
+            h = ( h << mu_norm ) | r;
+        }
+
+        return h;
+    }
+
+    // Calculate 2D Hilbert index given an ( x, y ) coordinate and the size of
+    // the hilbert space ( n )
+    template <typename I0, typename I1>
+    KOKKOS_INLINE_FUNCTION size_t Hilbert3D( I0 const &i0, I1 const &i1 ) const
+    {
+        int n = ( orig_dim.N0 > orig_dim.N1 ) ? orig_dim.N0 : orig_dim.N1;
+        int rx = 0;
+        int ry = 0;
+        int s = 0;
+        int d = 0;
+
+        I0 x = i0;
+        I1 y = i1;
+
+        // Split
+        for ( s = n / 2; s > 0; s /= 2 )
+        {
+            rx = ( x & s ) > 0;
+            ry = ( y & s ) > 0;
+            d += s * s * ( ( 3 * rx ) ^ ry );
+
+            // Rotate
+            if ( ry == 0 )
+            {
+                if ( rx == 1 )
+                {
+                    x = n - 1 - x;
+                    y = n - 1 - y;
+                }
+
+                int t = x;
+                x = y;
+                y = t;
+            }
+        }
+
+        // Return hilbert index
+        return d;
+    }
+
+    // rank 1
+    template <typename I0>
+    KOKKOS_INLINE_FUNCTION constexpr size_type operator()( I0 const &i0 ) const
+    {
+        // Calculate 3D hilbert index
+        size_t hilbert_index = hilbert3d( i0 + m_off.N0, 0, 0 ) - offset;
+
+        // Return hilbert index to map to 1 dimension
+        return hilbert_index;
+    }
+
+    // rank 2
+    template <typename I0, typename I1>
+    KOKKOS_INLINE_FUNCTION constexpr size_type operator()( I0 const &i0,
+                                                           I1 const &i1 ) const
+    {
+        // Calculate 3D hilbert index
+        size_t hilbert_index =
+            hilbert3d( i0 + m_off.N0, i1 + m_off.N1, 0 ) - offset;
+
+        // Return hilbert index to map to 2 dimensions
+        return hilbert_index;
+    }
+
+    // rank 3
+    template <typename I0, typename I1, typename I2>
+    KOKKOS_INLINE_FUNCTION constexpr size_type
+    operator()( I0 const &i0, I1 const &i1, I2 const &i2 ) const
+    {
+        // Calculate 3D hilbert index
+        size_t hilbert_index =
+            hilbert3d( i0 + m_off.N0, i1 + m_off.N1, i2 + m_off.N2 ) - offset;
+
+        // Use hilbert index to map to 3 dimensions
+        return hilbert_index;
+    }
+
+    // rank 4
+    template <typename I0, typename I1, typename I2, typename I3>
+    KOKKOS_INLINE_FUNCTION constexpr size_type
+    operator()( I0 const &i0, I1 const &i1, I2 const &i2, I3 const &i3 ) const
+    {
+        // Calculate 3D hilbert index
+        size_t hilbert_index =
+            hilbert3d( i0 + m_off.N0, i1 + m_off.N1, i2 + m_off.N2 ) - offset;
```

Comment:
It could be - my thinking was that logically it is easier to read as calculating the 3d hilbert index and then applying the offset as two logically separate steps. Would there be a computational advantage to placing it inside the hilbert3d function call and/or would you still recommend it otherwise?